server/core_benchmarks_test.go

```diff
@@ -249,3 +251,317 @@ func BenchmarkCoreTLSFanOut(b *testing.B) {
+	messageSizeCases := []int64{
+		100,        // 100B
+		1024,       // 1KiB
+		10240,      // 10KiB
+		512 * 1024, // 512KiB
+	}
+	numPubsCases := []int{
+		3,
+		5,
+		10,
+	}
+
+	// Custom error handler that ignores ErrSlowConsumer.
+	// Lots of them are expected in this benchmark which indiscriminately publishes at a rate higher
+	// than what the server can relay to subscribers.
+	ignoreSlowConsumerErrorHandler := func(conn *nats.Conn, s *nats.Subscription, err error) {
+		if errors.Is(err, nats.ErrSlowConsumer) {
+			// Swallow this error
+		} else {
+			_, _ = fmt.Fprintf(os.Stderr, "Warning: %s\n", err)
+		}
+	}
+
+	for _, messageSize := range messageSizeCases {
+		b.Run(
+			fmt.Sprintf("msgSz=%db", messageSize),
+			func(b *testing.B) {
+				for _, numPubs := range numPubsCases {
+					b.Run(
+						fmt.Sprintf("pubs=%d", numPubs),
+						func(b *testing.B) {
+
+							// Start server
+							defaultOpts := DefaultOptions()
+							server := RunServer(defaultOpts)
+							defer server.Shutdown()
+
+							opts := []nats.Option{
+								nats.MaxReconnects(-1),
+								nats.ReconnectWait(0),
+								nats.ErrorHandler(ignoreSlowConsumerErrorHandler),
+							}
+
+							clientUrl := server.ClientURL()
+
+							// start subscriber
+							ncSub, err := nats.Connect(clientUrl, opts...)
+							if err != nil {
+								b.Fatal(err)
+							}
+							defer ncSub.Close()
+
+							// track publishing errors
+							errors := make([]int, numPubs)
+							// track messages received by each publisher
+							counters := make([]int, numPubs)
+							// quit signals for each publisher
+							quitChs := make([]chan bool, numPubs)
+							for i := range quitChs {
+								quitChs[i] = make(chan bool, 1)
+							}
+
+							// TODO rename
+							completedPublishersCount := 0
+
+							var benchCompleteWg sync.WaitGroup
+							benchCompleteWg.Add(1)
+
+							ncSub.Subscribe(fmt.Sprintf("%s.*", subjectBaseName), func(msg *nats.Msg) {
+								// split subject by "." and get the publisher id
+								pubIdx, err := strconv.Atoi(strings.Split(msg.Subject, ".")[1])
+								if err != nil {
+									b.Fatal(err)
+								}
+
+								counters[pubIdx] += 1
+								if counters[pubIdx] == b.N {
+									completedPublishersCount++
+									if completedPublishersCount == numPubs {
+										benchCompleteWg.Done()
+									}
+								}
+							})
+
+							// random bytes as payload
+							messageData := make([]byte, messageSize)
+							rand.Read(messageData)
+
+							var publishersReadyWg sync.WaitGroup
+							// waits for all publishers sub-routines and for main thread to be ready
+							publishersReadyWg.Add(numPubs + 1)
+
+							// wait group to ensure all publishers have been torn down
+							var finishedPublishersWg sync.WaitGroup
+							finishedPublishersWg.Add(numPubs)
+
+							// create N publishers
+							for i := 0; i < numPubs; i++ {
+								// create publisher connection
+								ncPub, err := nats.Connect(clientUrl, opts...)
```

Comment:
Would have pulled all `Connect` into a separate for loop, and only move on to fork go subroutines after all connections are successfully established